DFE-Digital/teaching-vacancies
# ]Spike - Organisation jobseeker profile search

## Jira ticket URL

https://dfedigital.atlassian.net/browse/TEVA-4404

## Findings:

To search for jobseekers whose location preference area encompasses a given organisation's geopoint, we must first create a location preference area and store this in a jobseeker profile record. We can do this by leveraging the PostGIS Postgres extension and the RGeo gem. We must generate this polygon using both the location and radius - this location can be a point or one of our location polygons. 

In this spike, I've done this in a `before_save` callback in the JobseekerProfile model. When the location or radius changes, the location_preference is set. Perhaps this could be done in the jobseeker profile preference form model in the future instead.

We can then search for location preference areas which encompass a given organisation's geopoint by following the pattern set out for vacancy search. When a hiring staff user clicks on "Candidate profiles" below, in the relevant controller we can call Search::JobseekerProfileSearch.new(organisation).jobseekers (or something similar) to retrieve every jobseeker profile.

<img width="1060" alt="Screenshot 2022-11-01 at 11 58 03" src="https://user-images.githubusercontent.com/30624173/199227603-f6e1831e-ada6-4a39-af19-ef05aeb38480.png">

## To test

In the rails console, create the following jobseeker profiles:

```
JobseekerProfile.create(location: "E1 1BB", radius: 100)
JobseekerProfile.create(location: "London", radius: 50)
JobseekerProfile.create(location: "Manchester", radius: 50)
JobseekerProfile.create(location: "M1 2WD", radius: 50)
```

Then, grab Bexleyheath Academy:

`organisation = Organisation.find_by_name("Bexleyheath Academy")`

Two of these jobseeker profiles should now have location preference areas that cover Bexleyheath Academy (it is within 50 miles of the London polygon and E1 1BB). Therefore the following should retrieve two jobseeker profiles. 

Search::JobseekerProfileSearch.new(organisation).jobseekers
